python/lsst/pipe/base/execution_reports.py

```diff
@@ -0,0 +1,438 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+from __future__ import annotations
+
+__all__ = (
+    "QuantumGraphExecutionReport",
+    "TaskExecutionReport",
+    "DatasetTypeExecutionReport",
+    "lookup_quantum_data_id",
+)
+
+import dataclasses
+import itertools
+import logging
+import uuid
+from collections.abc import Iterable, Mapping
+from typing import Any
+
+import networkx
+import yaml
+from lsst.daf.butler import Butler, DataCoordinate, DatasetRef
+from lsst.resources import ResourcePathExpression
+
+from .graph import QuantumGraph, QuantumNode
+from .pipeline import PipelineDatasetTypes
+
+
+@dataclasses.dataclass
+class DatasetTypeExecutionReport:
+    """A report on the number of produced datasets as well as the status of
+    missing datasets based on metadata.
+
+    A DatasetTypeExecutionReport is created for each DatasetType in a
+    TaskExecutionReport.
+
+    See Also
+    --------
+    TaskExecutionReport
+    QuantumGraphExecutionReport
+    """
+
+    missing_failed: set[DatasetRef] = dataclasses.field(default_factory=set)
+    """Datasets not produced because their quanta failed directly in this
+    run (`set`).
+    """
+    missing_not_produced: dict[DatasetRef, bool] = dataclasses.field(default_factory=dict)
+    """Missing datasets which were not produced due either missing inputs or a
+    failure in finding inputs (`dict`).
+        bool: were predicted inputs produced?
+    """
+    missing_upstream_failed: set[DatasetRef] = dataclasses.field(default_factory=set)
+    """Datasets not produced due to an upstream failure (`set`).
+    """
+    n_produced: int = 0
+    """Count of datasets produced (`int`).
+    """
+
+    def to_summary_dict(self) -> dict[str, Any]:
+        """Summarize the DatasetTypeExecutionReport in a dictionary.
+
+        Returns
+        -------
+        A count of the DatasetTypes with each outcome; the number of produced,
+        missing_failed, missing_not_produced, and missing_upstream_failed
+        DatasetTypes. See above for attribute descriptions.
+        """
+        return {
+            "produced": self.n_produced,
+            "missing_failed": len(self.missing_failed),
+            "missing_not_produced": len(self.missing_not_produced),
+            "missing_upstream_failed": len(self.missing_upstream_failed),
+        }
+
+    def handle_missing_dataset(
+        self, output_ref: DatasetRef, failed: bool, status_graph: networkx.DiGraph
+    ) -> None:
+        """Sort missing datasets into outcomes.
+
+        Parameters
+        ----------
+        output_ref: `DatasetRef`
```

Comment:
Might need a space before the `:` here (and throughout the file).  At least that's how I usually see it, and I'm not sure if Sphinx cares or not.

Reply:
I'm pretty sure that it also has to be:
```
`~lsst.daf.butler.DatasetRef`
```
because `DatasetRef` is not a pipe_base class.